sdk/cosmos/azure-cosmos-spark_3-4_2-12/src/main/scala/com/azure/cosmos/spark/SparkInternalsBridge.scala

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.cosmos.spark
+
+import com.azure.cosmos.implementation.guava25.base.MoreObjects.firstNonNull
+import com.azure.cosmos.implementation.guava25.base.Strings.emptyToNull
+import com.azure.cosmos.spark.diagnostics.BasicLoggingTrait
+import org.apache.spark.TaskContext
+import org.apache.spark.sql.execution.metric.SQLMetric
+import org.apache.spark.util.AccumulatorV2
+
+import java.util.Locale
+import java.util.concurrent.atomic.AtomicBoolean
+
+object SparkInternalsBridge extends BasicLoggingTrait {
+  private val SPARK_REFLECTION_ACCESS_ALLOWED_PROPERTY = "COSMOS.SPARK_REFLECTION_ACCESS_ALLOWED"
+  private val SPARK_REFLECTION_ACCESS_ALLOWED_VARIABLE = "COSMOS_SPARK_REFLECTION_ACCESS_ALLOWED"
+
+  private val DEFAULT_SPARK_REFLECTION_ACCESS_ALLOWED = true
+
+  def getSparkReflectionAccessAllowed: Boolean = {
+    val allowedText = System.getProperty(
+      SPARK_REFLECTION_ACCESS_ALLOWED_PROPERTY,
+      firstNonNull(
+        emptyToNull(System.getenv.get(SPARK_REFLECTION_ACCESS_ALLOWED_VARIABLE)),
+        String.valueOf(DEFAULT_SPARK_REFLECTION_ACCESS_ALLOWED)))
+
+    try {
+      java.lang.Boolean.valueOf(allowedText.toUpperCase(Locale.ROOT))
+    }
+    catch {
+      case e: Exception =>
+        logError(s"Parsing spark reflection access allowed ${allowedText} failed. Using the default ${DEFAULT_SPARK_REFLECTION_ACCESS_ALLOWED}.", e)
+        DEFAULT_SPARK_REFLECTION_ACCESS_ALLOWED
+    }
+  }
+
+  private final lazy val reflectionAccessAllowed = new AtomicBoolean(getSparkReflectionAccessAllowed)
+
+  def getInternalCustomTaskMetricsAsSQLMetric(knownCosmosMetricNames: Set[String]) : Map[String, SQLMetric] = {
+
+    if (!reflectionAccessAllowed.get) {
+      Map.empty[String, SQLMetric]
+    } else {
+      Option.apply(TaskContext.get()) match {
+        case Some(taskCtx) => getInternalCustomTaskMetricsAsSQLMetricInternal(knownCosmosMetricNames, taskCtx)
+        case None => Map.empty[String, SQLMetric]
+      }
+    }
+  }
+
+  private def getAccumulators(taskCtx: TaskContext): Seq[AccumulatorV2[_, _]] = {
+    val taskMetrics:Object = taskCtx.taskMetrics()
+    taskMetrics.getClass.getMethod("accumulators").invoke(taskMetrics).asInstanceOf[Seq[AccumulatorV2[_, _]]]
```

Comment:
do we need to make the accumulators method accessible?` (accumulators.setAccessible(true))`

Reply:
But I guess we should - I also want to cache the reflection logic

Reply:
Not ready for review yet - forgot to set the draft mode